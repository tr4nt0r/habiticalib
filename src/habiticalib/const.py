"""Constants for Habiticalib."""

__version__ = "0.3.0"

DEFAULT_URL = "https://habitica.com/"
ASSETS_URL = "https://habitica-assets.s3.amazonaws.com/mobileApp/images/"

DEVELOPER_ID = "4c4ca53f-c059-4ffa-966e-9d29dd405daf"

BACKER_ONLY_GEAR = {
    "armor_special_ks2019": "BackerOnly-Equip-MythicGryphonArmor.gif",
    "eyewear_special_ks2019": "BackerOnly-Equip-MythicGryphonVisor.gif",
    "head_special_ks2019": "BackerOnly-Equip-MythicGryphonHelm.gif",
    "shield_special_ks2019": "BackerOnly-Equip-MythicGryphonShield.gif",
    "weapon_special_ks2019": "BackerOnly-Equip-MythicGryphonGlaive.gif",
}

PAGE_LIMIT = 60
