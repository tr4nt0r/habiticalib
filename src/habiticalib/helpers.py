"""Helper functions for Habiticalib."""

from dataclasses import asdict, is_dataclass
from datetime import date, datetime
from enum import Enum
import platform
from typing import Any
import uuid

import aiohttp

from .const import DEVELOPER_ID, __version__
from .types import HabiticaUserResponse, UserData, UserStyles


def join_fields(user_fields: list[str] | str) -> str:
    """Join user fields into a comma-separated string.

    This method takes a list of user fields or a single string and
    returns a string. If a list is provided, it joins the elements
    with a comma separator. If a string is provided, it returns
    the string as-is.

    Parameters
    ----------
    user_fields : list of str or str
        A list of user field strings or a single user field string.

    Returns
    -------
    str
        A comma-separated string of user fields if `user_fields` is a list,
        or the original string if `user_fields` is a single string.
    """
    return ",".join(user_fields) if isinstance(user_fields, list) else str(user_fields)


def get_user_agent() -> str:
    """Generate User-Agent string.

    The User-Agent string contains details about the operating system,
    its version, architecture, the habiticalib version, aiohttp version,
    and Python version.

    Returns
    -------
    str
        A User-Agent string with OS details, library versions, and a project URL.

    Examples
    --------
    >>> client.get_user_agent()
    'Habiticalib/0.0.0 (Windows 11 (10.0.22000); 64bit)
     aiohttp/3.10.9 Python/3.12.7  +https://github.com/tr4nt0r/habiticalib)'
    """
    os_name = platform.system()
    os_version = platform.version()
    os_release = platform.release()
    arch, _ = platform.architecture()
    os_info = f"{os_name} {os_release} ({os_version}); {arch}"

    return (
        f"Habiticalib/{__version__} ({os_info}) "
        f"aiohttp/{aiohttp.__version__} Python/{platform.python_version()} "
        " +https://github.com/tr4nt0r/habiticalib)"
    )


def get_x_client(x_client: str | None = None) -> str:
    """Generate the x-client header string.

    If a valid `x_client` is provided, it validates the User ID (UUID format).
    If no `x_client` is provided, it generates a default x-client header string.

    Parameters
    ----------
    x_client : str or None, optional
        A client identifier containing a UUID (first 36 characters) or None. If
        None, a default x-client header will be generated.

    Returns
    -------
    str
        The validated x-client header string or habiticalib's default x-client header.

    Raises
    ------
    ValueError
        If the provided `x_client` is not in a valid UUID format.

    Examples
    --------
    >>> get_x_client("123e4567-e89b-12d3-a456-426614174000 - MyHabiticaApp")
    '123e4567-e89b-12d3-a456-426614174000 - MyHabiticaApp'

    >>> get_x_client()
    '4c4ca53f-c059-4ffa-966e-9d29dd405daf - Habiticalib/0.0.0'
    """
    if x_client:
        try:
            uuid.UUID(x_client[:36])
        except ValueError as e:
            msg = (
                "Invalid User ID provided in x-client. Expected a valid "
                "UUID format. Please ensure the User ID is correct"
            )
            raise ValueError(msg) from e

        return x_client

    return f"{DEVELOPER_ID} - Habiticalib/{__version__}"


def extract_user_styles(user_data: HabiticaUserResponse) -> UserStyles:
    """Extract user styles from a user data object."""
    data: UserData = user_data.data
    return UserStyles.from_dict(asdict(data))


def deserialize_task(value: Any) -> Any:  # noqa: PLR0911
    """Recursively convert Enums to values, dates to ISO strings, UUIDs to strings."""

    if is_dataclass(value) and not isinstance(value, type):
        # Convert dataclass to dict and recursively deserialize
        return deserialize_task(asdict(value))
    if isinstance(value, Enum):
        return value.value  # Convert Enum to its value
    if isinstance(value, uuid.UUID):
        return str(value)  # Convert UUID to string
    if isinstance(value, datetime | date):
        return value.isoformat()  # Convert datetime/date to ISO string
    if isinstance(value, list):
        # Recursively apply deserialization to each item in the list
        return [deserialize_task(item) for item in value]
    if isinstance(value, dict):
        # Recursively apply deserialization to each key-value pair in the dictionary
        return {k: deserialize_task(v) for k, v in value.items()}
    return value  # Return other types unchanged
