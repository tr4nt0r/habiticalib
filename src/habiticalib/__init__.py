"""Modern asynchronous Python client library for the Habitica API."""

from .const import ASSETS_URL, DEFAULT_URL, __version__
from .exceptions import (
    BadRequestError,
    HabiticaException,
    NotAuthorizedError,
    NotFoundError,
    TooManyRequestsError,
)
from .helpers import deserialize_task
from .lib import Habitica
from .types import (
    Attributes,
    ChangeClassData,
    ContentData,
    Direction,
    Frequency,
    HabiticaClass,
    HabiticaClassSystemResponse,
    HabiticaContentResponse,
    HabiticaErrorResponse,
    HabiticaGroupMembersResponse,
    HabiticaLoginResponse,
    HabiticaQuestResponse,
    HabiticaResponse,
    HabiticaScoreResponse,
    HabiticaSleepResponse,
    HabiticaStatsResponse,
    HabiticaTagResponse,
    HabiticaTagsResponse,
    HabiticaTaskOrderResponse,
    HabiticaTaskResponse,
    HabiticaTasksResponse,
    HabiticaUserAnonymized,
    HabiticaUserExport,
    HabiticaUserResponse,
    Language,
    LoginData,
    QuestData,
    ScoreData,
    Skill,
    StatsUser,
    TagsUser,
    Task,
    TaskData,
    TaskFilter,
    TaskPriority,
    TaskType,
    UserAnonymizedData,
    UserData,
    UserStyles,
)

__all__ = [
    "ASSETS_URL",
    "DEFAULT_URL",
    "Attributes",
    "BadRequestError",
    "ChangeClassData",
    "ContentData",
    "Direction",
    "Frequency",
    "Habitica",
    "HabiticaClass",
    "HabiticaClassSystemResponse",
    "HabiticaContentResponse",
    "HabiticaErrorResponse",
    "HabiticaException",
    "HabiticaGroupMembersResponse",
    "HabiticaLoginResponse",
    "HabiticaQuestResponse",
    "HabiticaResponse",
    "HabiticaScoreResponse",
    "HabiticaSleepResponse",
    "HabiticaStatsResponse",
    "HabiticaTagResponse",
    "HabiticaTagsResponse",
    "HabiticaTaskOrderResponse",
    "HabiticaTaskResponse",
    "HabiticaTasksResponse",
    "HabiticaUserAnonymized",
    "HabiticaUserExport",
    "HabiticaUserResponse",
    "Language",
    "LoginData",
    "NotAuthorizedError",
    "NotFoundError",
    "QuestData",
    "ScoreData",
    "Skill",
    "StatsUser",
    "TagsUser",
    "Task",
    "TaskData",
    "TaskFilter",
    "TaskPriority",
    "TaskType",
    "TooManyRequestsError",
    "UserAnonymizedData",
    "UserData",
    "UserStyles",
    "__version__",
    "deserialize_task",
]
