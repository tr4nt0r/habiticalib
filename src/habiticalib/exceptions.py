"""Exceptions for Habiticalib."""

from datetime import datetime
from typing import Self

from multidict import CIMultiDictProxy

from habiticalib.types import HabiticaErrorResponse


class HabiticaException(Exception):  # noqa: N818
    """Base class for Habitica errors."""

    def __init__(
        self: Self,
        error: HabiticaErrorResponse,
        headers: CIMultiDictProxy,
    ) -> None:
        """Initialize the Exception."""
        self.error = error
        self.rate_limit: int | None = (
            int(r) if (r := headers.get("x-ratelimit-limit")) else None
        )
        self.rate_limit_remaining: int | None = (
            int(r) if (r := headers.get("x-ratelimit-remaining")) else None
        )
        self.rate_limit_reset: datetime | None = (
            datetime.strptime(r[:33], "%a %b %d %Y %H:%M:%S %Z%z")
            if (r := headers.get("x-ratelimit-reset"))
            else None
        )
        self.retry_after: int = round(float(headers.get("retry-after", 0)))

        super().__init__(error.message)


class NotAuthorizedError(HabiticaException):
    """NotAuthorized error."""


class NotFoundError(HabiticaException):
    """NotFound error."""


class BadRequestError(HabiticaException):
    """BadRequest error."""


class TooManyRequestsError(HabiticaException):
    """TooManyRequests error."""
