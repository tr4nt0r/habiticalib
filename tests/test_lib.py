"""Tests for user methods of Habiticalib."""

from aiohttp import ClientSession
from aioresponses import aioresponses
from syrupy.assertion import SnapshotAssertion

from habiticalib import Habitica

from .conftest import load_fixture


async def test_login(mock_aiohttp: aioresponses, snapshot: SnapshotAssertion) -> None:
    """Test login."""
    mock_aiohttp.post(
        "https://habitica.com/api/v3/user/auth/local/login",
        body=load_fixture("login.json"),
    )
    async with ClientSession() as session:
        habitica = Habitica(session, "test", "test")
        response = await habitica.login("test-username", "test-password")
        assert response == snapshot


async def test_user(mock_aiohttp: aioresponses, snapshot: SnapshotAssertion) -> None:
    """Test default user agent is set."""
    mock_aiohttp.get("https://habitica.com/api/v3/user", body=load_fixture("user.json"))
    async with ClientSession() as session:
        habitica = Habitica(session, "test", "test")
        response = await habitica.get_user()
        assert response == snapshot
