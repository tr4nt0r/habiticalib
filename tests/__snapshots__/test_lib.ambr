# serializer version: 1
# name: test_login
  HabiticaLoginResponse(data=LoginData(id=UUID('a380546a-94be-4b8e-8a0b-23e0d5c03303'), apiToken='cd0e5985-17de-4b4f-849e-5d506c5e4382', newUser=False, username='test-username', passwordResetCode=None), success=True, notifications=[], userV=None, appVersion='5.30.0')
# ---
# name: test_user
  HabiticaUserResponse(data=UserData(id=UUID('c18e1853-bded-47a9-82e2-adfdad08894d'), preferences=PreferencesUser(hair=HairPreferences(color='red', base=3, bangs=1, beard=0, mustache=0, flower=1), emailNotifications=EmailNotificationsPreferences(unsubscribeFromAll=False, newPM=True, kickedGroup=True, wonChallenge=True, giftedGems=True, giftedSubscription=True, invitedParty=True, invitedGuild=True, questStarted=True, invitedQuest=True, importantAnnouncements=True, weeklyRecaps=True, onboarding=True, majorUpdates=True, subscriptionReminders=True, contentRelease=True), pushNotifications=PushNotificationsPreferences(unsubscribeFromAll=False, newPM=True, wonChallenge=True, giftedGems=True, giftedSubscription=True, invitedParty=True, invitedGuild=True, questStarted=True, invitedQuest=True, majorUpdates=True, mentionParty=True, mentionJoinedGuild=True, mentionUnjoinedGuild=True, partyActivity=True, contentRelease=True), suppressModals=SuppressModalsPreferences(levelUp=False, hatchPet=False, raisePet=False, streak=False), tasks=TasksPreferences(activeFilter=ActiveFilterTask(habit='all', daily='all', todo='remaining', reward='all'), groupByChallenge=False, confirmScoreNotes=False, mirrorGroupTasks=[]), dayStart=0, size='slim', hideHeader=False, skin='915533', shirt='blue', timezoneOffset=-120, sound='rosstavoTheme', chair='none', allocationMode='flat', autoEquip=True, costume=False, dateFormat='MM/dd/yyyy', sleep=False, stickyHeader=True, disableClasses=False, newTaskEdit=False, dailyDueDefaultView=False, advancedCollapsed=False, toolbarCollapsed=False, reverseChatOrder=False, developerMode=False, displayInviteToPartyWhenPartyIs1=True, background='violet', automaticAllocation=None, webhooks={}, improvementCategories=[], timezoneOffsetAtLastCron=None, language=<Language.DE: 'de'>), flags=FlagsUser(customizationsNotification=False, tour=TourFlags(intro=-1, classes=-1, stats=-1, tavern=-1, party=-1, guilds=-1, challenges=-1, market=-1, pets=-1, mounts=-1, hall=-1, equipment=-1, groupPlans=-1), showTour=True, tutorial=TutorialFlags(common=CommonTutorial(habits=True, dailies=True, todos=True, rewards=True, party=True, pets=True, gems=True, skills=True, classes=True, tavern=True, equipment=True, items=True, mounts=True, inbox=True, stats=True), ios=IosTutorial(addTask=False, editTask=False, deleteTask=False, filterTask=False, groupPets=False, inviteParty=False, reorderTask=False)), dropsEnabled=False, itemsEnabled=False, lastNewStuffRead='', rewrite=True, classSelected=False, rebirthEnabled=False, levelDrops={}, recaptureEmailsPhase=0, weeklyRecapEmailsPhase=0, lastWeeklyRecap=datetime.datetime(2024, 10, 19, 18, 43, 39, 782000, tzinfo=datetime.timezone.utc), communityGuidelinesAccepted=False, cronCount=0, welcomed=True, armoireEnabled=True, armoireOpened=False, armoireEmpty=False, cardReceived=False, warnedLowHealth=False, verifiedUsername=True, newStuff=False, thirdPartyTools=None, mathUpdates=None, lastFreeRebirth=None, chatRevoked=None, chatShadowMuted=None, lastWeeklyRecapDiscriminator=None, onboardingEmailsPhase=None), auth=AuthUser(local=LocalAuth(email='test@example.com', username='test', lowerCaseUsername='test', has_password=True), timestamps=LocalTimestamps(created=datetime.datetime(2024, 10, 19, 18, 43, 39, 782000, tzinfo=datetime.timezone.utc), loggedin=datetime.datetime(2024, 10, 19, 18, 43, 39, 782000, tzinfo=datetime.timezone.utc), updated=datetime.datetime(2024, 10, 19, 18, 44, 51, 37000, tzinfo=datetime.timezone.utc)), facebook={}, google={}, apple={}), achievements=AchievementsUser(ultimateGearSets=UltimateGearSetsAchievments(healer=False, wizard=False, rogue=False, warrior=False), streak=0, challenges=[], perfect=0, quests=QuestsAchievments(bewilder=None, burnout=None, stressbeast=None, harpy=None, atom3=None, vice3=None, vice1=None, gryphon=None, evilsanta2=None, evilsanta=None, dilatory_derby=None, dilatory=None, atom2=None, atom1=None, dysheartener=None), backToBasics=None, dustDevil=None, primedForPainting=None, completedTask=None, createdTask=None, fedPet=None, hatchedPet=None, purchasedEquipment=None, tickledPink=None, goodAsGold=None, boneCollector=None, seeingRed=None, violetsAreBlue=None, shadyCustomer=None, joinedGuild=None, joinedChallenge=None, partyUp=None), backer=BackerUser(tier=None, npc=None, tokensApplied=None), contributor=ContributorUser(contributions=None, level=None, text=None), permissions=PermissionsUser(fullAccess=None, news=None, userSupport=None, challengeAdmin=None, moderator=None, coupons=None), purchased=PurchasedUser(plan=PlanPurchased(consecutive=ConsecutivePlan(trinkets=0, gemCapExtra=0, offset=0, count=0), mysteryItems=[], gemsBought=0, extraMonths=0, dateUpdated=None, perkMonthCount=-1, quantity=1), txnCount=0, background={'violet': True}, shirt={}, hair={}, skin={}, ads=False, mobileChat=None), history=HistoryUser(todos=[], exp=[]), items=ItemsUser(gear=GearItems(equipped=EquippedGear(weapon='weapon_special_fall2024Warrior', armor='armor_special_fall2024Warrior', head='head_special_fall2024Warrior', shield='shield_special_fall2024Warrior', back='back_mystery_201402', headAccessory='headAccessory_special_pinkHeadband', eyewear='eyewear_special_pinkHalfMoon', body='body_mystery_202003'), costume=EquippedGear(weapon=None, armor='armor_base_0', head='head_base_0', shield='shield_base_0', back=None, headAccessory=None, eyewear=None, body=None), owned={'headAccessory_special_blackHeadband': True, 'headAccessory_special_blueHeadband': True, 'headAccessory_special_greenHeadband': True, 'headAccessory_special_pinkHeadband': True, 'headAccessory_special_redHeadband': True, 'headAccessory_special_whiteHeadband': True, 'headAccessory_special_yellowHeadband': True, 'eyewear_special_blackTopFrame': True, 'eyewear_special_blueTopFrame': True, 'eyewear_special_greenTopFrame': True, 'eyewear_special_pinkTopFrame': True, 'eyewear_special_redTopFrame': True, 'eyewear_special_whiteTopFrame': True, 'eyewear_special_yellowTopFrame': True, 'eyewear_special_blackHalfMoon': True, 'eyewear_special_blueHalfMoon': True, 'eyewear_special_greenHalfMoon': True, 'eyewear_special_pinkHalfMoon': True, 'eyewear_special_redHalfMoon': True, 'eyewear_special_whiteHalfMoon': True, 'eyewear_special_yellowHalfMoon': True}), special=SpecialItems(birthdayReceived=[], birthday=0, thankyouReceived=[], thankyou=0, greetingReceived=[], greeting=0, nyeReceived=[], nye=0, valentineReceived=[], valentine=0, seafoam=0, shinySeed=0, spookySparkles=0, snowball=0, congrats=0, congratsReceived=[], getwell=0, getwellReceived=[], goodluck=0, goodluckReceived=[]), lastDrop=LastDropItems(count=0, date=datetime.datetime(2024, 10, 19, 18, 43, 39, 784000, tzinfo=datetime.timezone.utc)), currentMount='Velociraptor-Base', currentPet='Rat-Shade', quests={'dustbunnies': 1}, mounts={}, food={}, hatchingPotions={}, eggs={}, pets={}), invitations=InvitationsUser(party={}, guilds=[], parties=[]), party=PartyUser(quest=QuestParty(progress=ProgressQuest(up=0.0, down=0.0, collect={}, collectedItems=0), RSVPNeeded=False, key=None, completed=None), order='level', orderAscending='ascending', _id=None), profile=ProfileUser(blurb=None, imageUrl=None, name='test'), stats=StatsUser(buffs=BuffsStats(Str=0, per=0, con=0, stealth=0, streaks=False, seafoam=False, shinySeed=False, snowball=False, spookySparkles=False, Int=0), training=TrainingStats(Str=0.0, per=0, con=0, Int=0), hp=50.0, mp=10.0, exp=0, gp=0.0, lvl=1, Class=<HabiticaClass.WARRIOR: 'warrior'>, points=0, Str=0, con=0, per=0, toNextLevel=25, maxHealth=50, maxMP=30, Int=0), notifications=[], tags=[TagsUser(id=UUID('5358c71f-fe0f-4583-84a3-109b70c699fd'), name='Arbeit', challenge=None, group=None), TagsUser(id=UUID('45870b8b-3e1e-4e13-a722-6afcacdf689f'), name='Training', challenge=None, group=None), TagsUser(id=UUID('3e49b9db-0cc5-4070-b860-e16f1e4806a4'), name='Gesundheit + Wohlbefinden', challenge=None, group=None), TagsUser(id=UUID('0d85319e-7b9c-43e2-81ed-2a706521d5e3'), name='Schule', challenge=None, group=None), TagsUser(id=UUID('b7cf86bf-97ac-42d3-b5d0-5ffe998b62b4'), name='Teams', challenge=None, group=None), TagsUser(id=UUID('c584d792-0a9b-4646-a6ca-a67bb6ef3c36'), name='Hausarbeiten', challenge=None, group=None), TagsUser(id=UUID('0602fd13-edf2-46a7-99c0-c5084046efbb'), name='Kreativit√§t', challenge=None, group=None)], inbox=InboxUser(newMessages=0, optOut=False, blocks=[], messages={}), tasksOrder=TasksOrderUser(habits=[UUID('21b10675-e238-4462-be11-1a4f8012fc9c'), UUID('69909140-a920-4b83-b018-b9156760aca3'), UUID('7e0679cf-e9cd-4604-b429-d061ea16ce72'), UUID('270df162-d47d-488b-9c1d-4fc4f0e2b2d2'), UUID('80ad00b5-622e-4d1a-af4d-6a6c199f3571'), UUID('0712669d-8374-4211-84ea-715cbcbae9c1'), UUID('56dee98d-038a-43f3-a514-59b1ed9a52ee'), UUID('25df4765-754c-47b1-a011-83b31f70c4fb')], dailys=[UUID('e22e9ef0-15be-478a-81c7-f9fdbb4fac10'), UUID('2a74baa2-3acd-426c-bce4-6b4bb145a8d5'), UUID('e18d0899-7c59-41c0-affd-318b8850c3e5'), UUID('3d0cecac-73c1-462b-ab39-a39cea107e6b'), UUID('1ed21cd0-e6f5-4707-b1b4-fb376d462387'), UUID('9117ad5e-2cda-43e0-a07c-0b175d32c0eb')], todos=[UUID('ac13c62c-9375-4809-a357-0e42b01a1b43'), UUID('2388ccad-0387-4ea6-9968-2d10d994e903'), UUID('5c6efb91-6ae3-4cad-8b19-2dd2fc52a965'), UUID('80afe0b9-e367-4473-8117-b8bc7115f54a'), UUID('4cc21c36-886c-4a43-b580-877e958a37fd'), UUID('10a45996-7133-4ca1-84d8-abf1093d9dcc'), UUID('fb867459-ae50-4cee-9017-c13206c6dfa2')], rewards=[UUID('4b18ccf6-3934-43a5-a153-31af8b2a36d9')]), extra={}, pushDevices=[], webhooks=[], loginIncentives=0, invitesSent=0, pinnedItems=[PinnedItemsUser(path='gear.flat.weapon_warrior_0', Type='marketGear'), PinnedItemsUser(path='gear.flat.armor_warrior_1', Type='marketGear'), PinnedItemsUser(path='gear.flat.shield_warrior_1', Type='marketGear'), PinnedItemsUser(path='gear.flat.head_warrior_1', Type='marketGear'), PinnedItemsUser(path='potion', Type='potion'), PinnedItemsUser(path='armoire', Type='armoire')], pinnedItemsOrder=[], unpinnedItems=[], secret=None, balance=0.0, lastCron=datetime.datetime(2024, 10, 19, 18, 43, 39, 784000, tzinfo=datetime.timezone.utc), needsCron=False, challenges=[], guilds=[], newMessages={}), success=True, notifications=[], userV=5, appVersion='5.28.8')
# ---
